bindings/src/test/java/io/pravega/storage/extendeds3/S3Mock.java

```diff
@@ -0,0 +1,315 @@
+import com.emc.object.s3.request.CopyPartRequest;
+import com.emc.object.s3.request.DeleteObjectsRequest;
+import com.emc.object.s3.request.PutObjectRequest;
+import com.emc.object.s3.request.SetObjectAclRequest;
+import io.pravega.common.io.StreamHelpers;
+import io.pravega.common.util.BufferView;
+import io.pravega.common.util.ByteArraySegment;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
+import lombok.AllArgsConstructor;
+import lombok.val;
+
+/**
+ * In-memory mock for S3.
+ */
+public class S3Mock {
+    //region Private
+
+    @GuardedBy("objects")
+    private final Map<String, Map<Integer, CopyPartRequest>> multipartUploads;
+    @GuardedBy("objects")
+    private final Map<String, ObjectData> objects;
+
+    //endregion
+
+    //region Constructor
+
+    public S3Mock() {
+        this.objects = new HashMap<>();
+        this.multipartUploads = new HashMap<>();
+    }
+
+    //endregion
+
+    //region Mock Implementation
+
+    private String getObjectName(String bucketName, String key) {
+        return String.format("%s-%s", bucketName, key);
+    }
+
+    public PutObjectResult putObject(PutObjectRequest request) {
+        if (request.getObjectMetadata() != null) {
+            request.setObjectMetadata(null);
+        }
+        String objectName = getObjectName(request.getBucketName(), request.getKey());
+        synchronized (this.objects) {
+            if (this.objects.containsKey(objectName)) {
+                throw new S3Exception(String.format("Object '%s/%s' exists.", request.getBucketName(), request.getKey()), 0, "PreconditionFailed", "");
+            }
+
+            this.objects.put(objectName, new ObjectData(BufferView.empty(), request.getAcl()));
+            return new PutObjectResult();
+        }
+    }
+
+    public void putObject(String bucketName, String key, Range range, Object content) {
+        String objectName = getObjectName(bucketName, key);
+        synchronized (this.objects) {
+            ObjectData od = this.objects.get(objectName);
+            if (od == null) {
+                throw new S3Exception(String.format("Object '%s/%s' does not exist.", bucketName, key), 404, "NoSuchKey", key);
+            }
+
+            final int startOffset = (int) (long) range.getFirst();
+            final int length = (int) (range.getLast() + 1 - range.getFirst());
+            BufferView objectContent = od.content;
+            if (startOffset < objectContent.getLength()) {
+                objectContent = new ByteArraySegment(objectContent.slice(0, startOffset).getCopy());
+            } else if (startOffset > objectContent.getLength()) {
+                throw new S3Exception("", 0, "InvalidRange", "");
```

Comment:
http code for invalid range  is 416.
Let's use org.apache.http.HttpStatus.SC_REQUESTED_RANGE_NOT_SATISFIABLE

Reply:
what happens when requested length is more than length of object?

Reply:
HTTP code - see above.

Length out of bounds - the `.slice` invocation above will throw.

Reply:
I think we do have that reference on line 236

Reply:
Duh. Fixed all of them.